dask_ndmeasure/__init__.py

```diff
@@ -1,9 +1,88 @@
 # -*- coding: utf-8 -*-
 
+from __future__ import division
+
+
 __author__ = """John Kirkham"""
 __email__ = "kirkhamj@janelia.hhmi.org"
 
 
 from ._version import get_versions
 __version__ = get_versions()['version']
 del get_versions
+
+
+import numpy
+import dask.array
+
+from . import _compat
+
+
+def center_of_mass(input, labels=None, index=None):
+    """
+    Calculate the center of mass of the values of an array at labels.
+
+    Parameters
+    ----------
+    input : ndarray
+        Data from which to calculate center-of-mass. The masses can either
+        be positive or negative.
+    labels : ndarray, optional
+        Labels for objects in `input`, as generated by `ndimage.label`.
+        Only used with `index`.  Dimensions must be the same as `input`.
+    index : int or sequence of ints, optional
+        Labels for which to calculate centers-of-mass. If not specified,
+        all labels greater than zero are used.  Only used with `labels`.
+
+    Returns
+    -------
+    center_of_mass : tuple, or list of tuples
+        Coordinates of centers-of-mass.
+    """
+
+    input = _compat._asarray(input)
+
+    if labels is None:
+        labels = (input != 0).astype(numpy.int64)
+        index = None
+
+    if index is None:
+        labels = (labels > 0).astype(numpy.int64)
+        index = dask.array.ones(tuple(), dtype=numpy.int64, chunks=tuple())
+
+    labels = _compat._asarray(labels)
+    index = _compat._asarray(index)
+
+    # SciPy transposes these for some reason.
+    # So we do the same thing here.
+    # This only matters if index is some array.
+    index = index.T
+
+    if input.shape != labels.shape:
+        raise ValueError("The input and labels arrays must be the same shape.")
+
+    input_ind = _compat._indices(input.shape, chunks=input.chunks)
+
+    lbl_mtch = (
+        index[(Ellipsis,) + labels.ndim * (None,)] == labels[index.ndim * (None,)]
+    )
+    lbl_mtch = lbl_mtch.astype(input.dtype)
```

Comment:
Probably the bulk of the code before this point can be refactored into some general purpose utility functions. It seems there are many functions in SciPy's `ndimage`'s `measurements` module that take an optional label array and optional indices.

In the end, we need to do the same thing in all cases, which is generate a set of masks that represent each label index we want to select. Then each function does something with these masks, which may be different in each case.